SpdLog/log_manager.h

```diff
@@ -0,0 +1,175 @@
+#pragma once
+#include <filesystem>
+#include <string>
+#include <iostream>
+#include <optional>
+
+#define SPDLOG_LEVEL_NAMES \
+{ \
+  std::string_view{"TRACE"}, \
+  std::string_view{"DEBUG"}, \
+  std::string_view{"INFO"}, \
+  std::string_view{"WARN"}, \
+  std::string_view{"ERROR"}, \
+  std::string_view{"FATAL"}, \
+  std::string_view{"OFF"} \
+};
+
+// SPDLOG_NO_EXCEPTIONS calls abort() not sure whether this is good idea...
+
+#define SPDLOG_NO_EXCEPTIONS
+
+#pragma warning(push)
+#pragma warning ( disable: 4307 ) // spdlog warning C4307: '*': integral constant overflow
+#include <spdlog/logger.h>
+#include <spdlog/spdlog.h>
+#include <spdlog/sinks/rotating_file_sink.h>
+#pragma warning(pop)
+
+class log_ptr;
+
+class log_manager
+{
+    static const std::string default_sink_name;
+    static const std::string messaging_sink_name;
+
+    static const std::string log_pattern;
+
+    log_manager();
+    std::unordered_map<std::string, spdlog::sink_ptr> sinks_;
+
+    std::shared_ptr<spdlog::logger> get_internal_logger(const std::string& sink_name, const std::string& logger_name);
+
+public:
+    friend log_ptr;
+
+    static log_ptr get_logger(std::string logger_name);
+    static log_ptr get_messaging_logger(std::string logger_name);
+
+    void initialize(const std::filesystem::path& base_path)
+    {
+        static constexpr std::size_t max_size = 1024;
+        static constexpr std::size_t max_files = 3;
+        static spdlog::level::level_enum global_level = spdlog::level::debug;
+
+        spdlog::drop_all(); // remove implicit default logger
+        spdlog::set_level(global_level);
+
+        spdlog::file_event_handlers handlers;
+        handlers.after_open = [](const spdlog::filename_t&, std::FILE *file_stream) { fputs("*** SpdLog Service v2023.4.0.1070 ***\n", file_stream); };
+
+        const auto default_log_path = base_path / "default.log";
+        auto default_log_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(default_log_path.generic_string(), max_size, max_files, false, handlers);
+        const auto default_logger = std::make_shared<spdlog::logger>("SpdLog", default_log_sink);
+        
+        spdlog::set_default_logger(default_logger);
+        spdlog::set_pattern(log_pattern); // must be set after set_default_logger
+
+        const auto messaging_log_path = base_path / "messaging.log";
+        const auto messaging_log_sink = std::make_shared<spdlog::sinks::rotating_file_sink_mt>(messaging_log_path.generic_string(), max_size, max_files, false, handlers);
+
+        sinks_[default_sink_name] = default_log_sink;
+        sinks_[messaging_sink_name] = messaging_log_sink;
+
+        spdlog::flush_every(std::chrono::seconds(1));
+    }
+
+    static void shutdown()
+    {
+        spdlog::shutdown();
+    }
+
+    static log_manager &instance();
+};
+
+class log_ptr
+{
+    std::string sink_name_;
+    std::string name_;
+    std::optional<std::shared_ptr<spdlog::logger>> logger_;

```

Comment:
optional<shared_ptr> ? sounds overcomplicated...
either shared_ptr should suffice, or weak_ptr if object lifespan is in question?

Reply:
well, I used `std::optional` so I can nullify it in ctor (because initial value is unset) but simple `std::shared_ptr` should do the same trick 233fde722b049443fea4613171529d4f728d9674